src/util/slice.rs

```diff
@@ -30,22 +31,24 @@ use std::slice;
 /// deallocated.
 #[derive(Clone, Eq)]
 pub struct Slice {
-    data: *const u8,
-    size: usize,
+    data: Bytes,
 }
 
 impl Slice {
     pub fn new(data: *const u8, size: usize) -> Self {
-        Self { data, size }
+        let data = if !data.is_null() {
+            let slice = unsafe { slice::from_raw_parts(data, size) };
+            Bytes::copy_from_slice(&slice[..])
```

Comment:
It seems we can't avoid the copying. And the original version just acquire a ptr.

Reply:
Yeah, it is unfortunate, though I am curious if we can just replace all usage of raw slices with bytes? I took a quick look and it seems that there are only a few places that its required.